Test Failure: JDBC41Test.testTransactionTimeoutAbort times out before enlistment

Test Failure: com.ibm.ws.jdbc.fat.v41.JDBC41Test.testTransactionTimeoutAbort

```
testTransactionTimeoutAbort:junit.framework.AssertionFailedError: 2021-04-25-22:14:34:232 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testTransactionTimeoutAbort on servlet jdbc.fat.v41.web.BasicTestServlet
java.sql.SQLException: addSync: caught Exception
	at com.ibm.ws.rsadapter.AdapterUtil.toSQLException(AdapterUtil.java:869)
	at com.ibm.ws.rsadapter.jdbc.WSJdbcConnection.activate(WSJdbcConnection.java:2332)
	at com.ibm.ws.rsadapter.jdbc.WSJdbcConnection.createStatement(WSJdbcConnection.java:1073)
	at com.ibm.ws.rsadapter.jdbc.WSJdbcConnection.createStatement(WSJdbcConnection.java:1059)
	at jdbc.fat.v41.web.BasicTestServlet.testTransactionTimeoutAbort(BasicTestServlet.java:1147)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at componenttest.app.FATServlet.doGet(FATServlet.java:71)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1011)
	at com.ibm.ws.webcontainer.servlet.CacheServletWrapper.handleRequest(CacheServletWrapper.java:75)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:938)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1159)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:428)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:387)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:566)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:500)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:360)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:327)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:504)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:574)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:958)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1047)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.resource.ResourceException: addSync: caught Exception
	at com.ibm.ejs.j2c.XATransactionWrapper.addSync(XATransactionWrapper.java:180)
	at com.ibm.ejs.j2c.ConnectionManager.initializeForUOW(ConnectionManager.java:973)
	at com.ibm.ejs.j2c.ConnectionManager.involveMCInTran(ConnectionManager.java:692)
	at com.ibm.ejs.j2c.ConnectionManager.associateConnection(ConnectionManager.java:1541)
	at com.ibm.ejs.j2c.ConnectionManager.associateConnection(ConnectionManager.java:1497)
	at com.ibm.ws.rsadapter.jdbc.WSJdbcConnection.activate(WSJdbcConnection.java:2324)
Caused by: javax.transaction.RollbackException: Transaction rolled back
	at com.ibm.tx.jta.impl.TransactionImpl.registerSynchronization(TransactionImpl.java:2284)
	at com.ibm.tx.jta.embeddable.impl.EmbeddableTranManagerSet.registerSynchronization(EmbeddableTranManagerSet.java:229)
	at com.ibm.tx.jta.embeddable.impl.EmbeddableTranManagerSet.registerSynchronization(EmbeddableTranManagerSet.java:216)
	at com.ibm.ws.transaction.services.TransactionManagerService.registerSynchronization(TransactionManagerService.java:406)
	at com.ibm.ejs.j2c.XATransactionWrapper.addSync(XATransactionWrapper.java:170)
". expected:<0> but was:<1>
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:537)
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:495)
	at componenttest.topology.utils.HttpUtils.findStringInUrl(HttpUtils.java:472)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:444)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:414)
	at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:83)
	at componenttest.custom.junit.runner.SyntheticServletTest.invokeExplosively(SyntheticServletTest.java:40)
	at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:197)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:318)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:171)
	at componenttest.rules.repeater.RepeatTests$CompositeRepeatTestActionStatement.evaluate(RepeatTests.java:115)
```

An 8 second delay (test infrastructure slowness) interferes with the timing logic of a test case which intentionally causes a transaction timeout in order to test that code path by reducing the transaction timeout to 3 seconds.  The problem with the test is that it expected to be able to quickly run a command to enlist in the transaction within those 3 seconds, but due to the 8 second delay, that is not possible and the enlistment attempt is rejected with an error that the test case isn't coded to expect.

--- a/dev/com.ibm.ws.jdbc_fat_v41/fat/src/com/ibm/ws/jdbc/fat/v41/JDBC41Test.java
+++ b/dev/com.ibm.ws.jdbc_fat_v41/fat/src/com/ibm/ws/jdbc/fat/v41/JDBC41Test.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2017,2020 IBM Corporation and others.
+ * Copyright (c) 2017,2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -47,9 +47,11 @@ public class JDBC41Test extends FATServletClient {
     @AfterClass
     public static void tearDown() throws Exception {
         server.stopServer("J2CA0081E.*slowDS", // expected by testNetworkTimeoutSimple
+                          "J2CA0026E.*RollbackException", // from testTransactionTimeoutAbort when transaction times out before enlist
                           "J2CA0027E", // expected by testTransactionTimeoutAbort - XAds removed because it does not appear on the same line as the message ID in every language
                           "DSRA0302E.*XAException*", // expected by testTransactionTimeoutAbort
                           "DSRA0304E.*XAException*", // expected by testTransactionTimeoutAbort
+                          "DSRA9400E.*addSync", // from testTransactionTimeoutAbort when transaction times out before enlist
                           "J2CA0079E"); // expected by testMBeanPurgeAbort
     }
 

--- a/dev/com.ibm.ws.jdbc_fat_v41/test-applications/basicfat/src/jdbc/fat/v41/web/BasicTestServlet.java
+++ b/dev/com.ibm.ws.jdbc_fat_v41/test-applications/basicfat/src/jdbc/fat/v41/web/BasicTestServlet.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2017 IBM Corporation and others.
+ * Copyright (c) 2017,2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -1137,16 +1137,29 @@ public class BasicTestServlet extends FATDatabaseServlet {
      */
     @Test
     @SkipIfDataSourceProperties({ SYBASE, INFORMIX_JDBC }) // no 4.1 sybase or ifx driver
-    @AllowedFFDC({ "javax.transaction.xa.XAException", "java.lang.NullPointerException", "oracle.jdbc.xa.OracleXAException",
+    @AllowedFFDC({
+                   "javax.resource.ResourceException", // times out before enlistment
+                   "javax.transaction.RollbackException", // times out before enlistment
+                   "javax.transaction.xa.XAException", "java.lang.NullPointerException", "oracle.jdbc.xa.OracleXAException",
                    "org.apache.derby.shared.common.sanity.AssertFailure" })
     public void testTransactionTimeoutAbort() throws Exception {
         Connection conn = xads.getConnection();
         tran.setTransactionTimeout(3);
         tran.begin();
 
-        Statement stmt = conn.createStatement();
-        stmt.execute("select * from " + colorTable);
-        stmt.close();
+        try {
+            Statement stmt = conn.createStatement();
+            stmt.execute("select * from " + colorTable);
+            stmt.close();
+        } catch (SQLException x) {
+            boolean isRollbackException = false;
+            for (Throwable cause = x; cause != null && !isRollbackException; cause = cause.getCause())
+                isRollbackException = cause instanceof RollbackException;
+            if (isRollbackException)
+                ; // transaction timed out before we could enlist in it
+            else
+                throw x;
+        }
 
         System.out.println("Wait up to 2 minutes for transaction to be marked for rollback");
         for (long start = System.nanoTime(); //
@@ -1169,7 +1182,7 @@ public class BasicTestServlet extends FATDatabaseServlet {
         conn = xads.getConnection();
         try {
             tran.begin();
-            stmt = conn.createStatement();
+            Statement stmt = conn.createStatement();
             stmt.execute("select * from " + colorTable);
             stmt.close();
             tran.commit();

